File: Coding_Questions/JavaScript/GenerateDocument.js
// ---Generate Document---

const generateDocument = (characters, document) => {
  if (characters.length < document.length) {
    return false;
  } else if (document === '') {
    return true;
  }

  const characterObject = {};

  for (const char of characters) {
    if (characterObject[char] === undefined) {
      characterObject[char] = 1;
    } else {
      characterObject[char]++;
    }
  }

  for (const char of document) {
    if (characterObject[char] !== undefined) {
      characterObject[char]--;
      if (characterObject[char] < 0) {
        return false;
      }
    } else {
      return false;
    }
  }

  return true;
};

const a = generateDocument(
  'Bste!hetsi ogEAxpelrt x ',
  'AlgoExpert is the Best!'
);
console.log(a); // true

File: Coding_Questions/Python/GenerateDocument.py
# ---Generate Document---

# solution 1
def generate_document(characters: str, document: str) -> bool:
    if len(characters) < len(document):
        return False
    elif document == "":
        return True

    characters_dict = dict()

    for char in characters:
        if char in characters_dict:
            characters_dict[char] += 1
        else:
            characters_dict[char] = 1

    for char in document:
        if char in characters_dict:
            characters_dict[char] -= 1
            if characters_dict[char] < 0:
                return False
        else:
            return False

    return True


a = generate_document("Bste!hetsi ogEAxpelrt x ", "AlgoExpert is the Best!")
assert a is True
